### ISSUE TITLE
Slow performance on TPC-DS tests

### ISSUE THREAD

**Issue by andrey-mindrin** [apache/iceberg#4217]
We’ve tested Iceberg performance vs Hive format by using Spark TPC-DS performance tests (scale factor 1000) from Databricks and found 50% less performance in Iceberg tables.

Environment:
1. On premises cluster which runs Spark 3.1.2 with Iceberg 0.13.0 with the same number executors, cores, memory, etc.
2. Parquet codec snappy
3. Tables were partitioned like in original Hive tables
4. Tables were COW and they were created in Spark from Hive tables with CTAS.

**Comment by RussellSpitzer** [comment]
Could you add some more details? Like what particular tests were slower? How did you configure the Iceberg Table? What kind of storage did you use?

It may be helpful to see the exact commands run to replicate the test.

**Comment by andrey-mindrin** [comment]
Used https://github.com/databricks/spark-sql-perf, tested on HDFS. Try for example q7-v2.4.

**Comment by andrey-mindrin** [comment]
Additional information. Total size of data is about 300 Gb in parquet, tried different options for iceberg tables including vectorization, distributed by hash and range. Iceberg tables created with a query like

`create table ice_table using iceberg partition by (column) location ‘hdfs://location’ tblproperties(‘vectorization-enabled’=true,’write.parquet.compression-codec’=’snappy’,’ write.distribution-mode’=’hash’) as select * from hive_table `

Analyzing query plans in Spark for query q7-v2.4 shows different join types – SortMergeJoin in Iceberg and BroadcastHashJoin in Hive tables.

**Comment by wypoon** [comment]
I have done some experiments along this line too. The way I created the Iceberg tables from the TPD-DS Hive tables is using the snapshot procedure to create them in a different database:
```
spark.sql("create database tpcds_iceberg")
spark.sql("use tpcds")
val tables = spark.sql("show tables")
tables.collect().map(r => r(1).toString).foreach(t =>
  spark.sql(s"call spark_catalog.system.snapshot('tpcds.$t', 'tpcds_iceberg.$t')")
)
```
This creates Iceberg tables backed by the same underlying data. As we're not writing to the tables, this does not create any problems.
For the same value of `spark.sql.autoBroadcastJoinThreshold` Spark will use a broadcast join in some of the TPC-DS queries on Hive tables but use a SortMergeJoin on the Iceberg tables. This is because the way Spark estimates the size of the relation for the native table case is different from the way Iceberg estimates the size. For the native table case, Spark uses file size in its estimation. File size can be a significant underestimate as in your case, where you are using Snappy-compressed Parquet files, as most folks don't even set `spark.sql.sources.fileCompressionFactor` (which defaults to 1.0). Iceberg estimates the size of the relation by multiplying the estimated width of the requested columns by the number of rows. In my original commit for https://github.com/apache/iceberg/pull/3038, I used the same approach to estimating the size of the relation that Spark uses for `FileScan`s, but @rdblue suggested to use the approach actually adopted.
To get Spark to use broadcast joins for the Iceberg tables, you can set a higher value for `spark.sql.autoBroadcastJoinThreshold`. The problem is that it is hard to compare apples with apples in this case.

**Comment by ConeyLiu** [comment]
We also have benchmarked iceberg with TPCDS. And got the following finds:
1. As @wypoon said, spark reading parquet and reading iceberg using different relation size estimation, which leads to the different table plan. Such as BroadcastJoin to SortMergeJoin.
2. Spark reading parquet with vectorized data reading by default. However, we have just enabled it by default in iceberg recently.  You could enable it by yourself. This could improve data reading a lot. And I have a pr (https://github.com/apache/iceberg/pull/3249) for optimizing iceberg parquet decimal vectorized reading.
3. Spark supports DDP for build-in datasource while it supports iceberg-like datasource since spark 3.2. This could influence the performance a lot from TPCDS side.

**Comment by andrey-mindrin** [comment]
I can confirm that iceberg performance on Spark 3.2 is similar to Hive tables. It seems the problem was in Spark 3.1 only.

**Comment by rdblue** [comment]
Also make sure that vectorization is enabled for tables. This is the default in master and the upcoming release, but older versions need to enable vectorization by setting `read.parquet.vectorization.enabled=true` in table properties.

**Comment by kbendick** [comment]
Along with data vectorizations, I would encourage people to use the following system properties which affect Arrows performance that are used within our tests (particularly if benchmarking as Arrow does duplicate some checks that we already do, which can have a big impact on hot code paths and the eventually compiled code):

https://github.com/apache/iceberg/blob/c9442f1c0f0524b7ca5ac3ccbb927cc80b764239/spark/v3.2/build.gradle#L104-L110

**Comment by aokolnychyi** [comment]
Resolved by PR #6550.
